Firebase/Database/Core/FServerValues.m

```diff
@@ -117,49 +183,54 @@ + (id)resolveComplexServerOp:(NSDictionary *)op
+    [write enumerateWrites:^(FPath *subPath, id<FNode> node, BOOL *stop) {
+      id<ValueProvider> existing =
+          [[DeferredValueProvider alloc] initWithSyncTree:tree
+                                                   atPath:[path child:subPath]];
       id<FNode> resolvedNode =
           [FServerValues resolveDeferredValueSnapshot:node
-                                         withExisting:existing
+                                      withJITExisting:existing
                                          serverValues:serverValues];
       // Node actually changed, use pointer inequality here
       if (resolvedNode != node) {
-          resolved = [resolved addWrite:resolvedNode atPath:path];
+          resolved = [resolved addWrite:resolvedNode atPath:subPath];
       }
     }];
     return resolved;
 }
 
-+ (id)resolveDeferredValueTree:(FSparseSnapshotTree *)tree
-                  withExisting:(id<FNode>)existing
-                  serverValues:(NSDictionary *)serverValues {
-    FSparseSnapshotTree *resolvedTree = [[FSparseSnapshotTree alloc] init];
-    [tree
-        forEachTreeAtPath:[FPath empty]
-                       do:^(FPath *path, id<FNode> node) {
-                         [resolvedTree
-                             rememberData:
-                                 [FServerValues
-                                     resolveDeferredValueSnapshot:node
-                                                     withExisting:
-                                                         [existing
-                                                             getChild:path]
-                                                     serverValues:serverValues]
-                                   onPath:path];
-                       }];
-    return resolvedTree;
++ (id<FNode>)resolveDeferredValueSnapshot:(id<FNode>)node
+                             withSyncTree:(FSyncTree *)tree
+                                   atPath:(FPath *)path
+                             serverValues:(NSDictionary *)serverValues {
+    id<ValueProvider> jitExisting =
+        [[DeferredValueProvider alloc] initWithSyncTree:tree atPath:path];
+    return [FServerValues resolveDeferredValueSnapshot:node
+                                       withJITExisting:jitExisting
+                                          serverValues:serverValues];
 }
 
 + (id<FNode>)resolveDeferredValueSnapshot:(id<FNode>)node
                              withExisting:(id<FNode>)existing
                              serverValues:(NSDictionary *)serverValues {
+    id<ValueProvider> jitExisting =
+        [[ExistingValueProvider alloc] initWithSnapshot:existing];
+    return [FServerValues resolveDeferredValueSnapshot:node
+                                       withJITExisting:jitExisting
+                                          serverValues:serverValues];
+}
+
++ (id<FNode>)resolveDeferredValueSnapshot:(id<FNode>)node
+                          withJITExisting:(id<ValueProvider>)existing
```

Comment:
Nit: Possibly rename to `withValueProvider`.

Reply:
🤦‍♂️You're right, of course. Guess I'll add another commit and wait another hour to check in.